### ISSUE TITLE
pympress fails to start on ARM macOS

### ISSUE THREAD

**Issue by DocEpsilon** [Cimbali/pympress#190]
### Environment
- Operating system and version: macOS Big Sur 11.2.3
- Arm64 Architecture (M1 MacBook Pro 13")
- Python version: 3.9

### Pympress
- version 1.5.3 as well as current git clone
- Installation method: ` brew install --build-from-source pympress `for 1.5.3 and` python3 setup.py install` for git clone

### Expected behaviour
What are you trying to achieve? What's the expected outcome?
- Trying to start pympress and display a pdf.
- Pympress fails to start with assertion error.

### Steps to reproduce
What are you doing?
`% pympress presentation.pdf    `

### Problem
What's happening instead of what you are expecting?
Expecting the pympress to start and the prensentation to be shown.
Instead: 
```
% pympress presentation.pdf

** (process:22380): WARNING **: 20:38:42.818: Failed to load shared library 'libpango-1.0.0.dylib' referenced by the typelib: dlopen(libpango-1.0.0.dylib, 9): image not found
Traceback (most recent call last):
  File "/opt/homebrew/bin/pympress", line 33, in <module>
    sys.exit(load_entry_point('pympress==1.5.3', 'gui_scripts', 'pympress')())
  File "/opt/homebrew/bin/pympress", line 25, in importlib_load_entry_point
    return next(matches).load()
  File "/opt/homebrew/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/importlib/metadata.py", line 77, in load
    module = import_module(match.group('module'))
  File "/opt/homebrew/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 790, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/opt/homebrew/Cellar/pympress/1.5.3_2/libexec/lib/python3.9/site-packages/pympress/__main__.py", line 79, in <module>
    from gi.repository import Gtk, Gdk, GLib
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 664, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 627, in _load_backward_compatible
  File "/opt/homebrew/opt/pygobject3/lib/python3.9/site-packages/gi/importer.py", line 144, in load_module
    importlib.import_module('gi.repository.' + dep.split("-")[0])
  File "/opt/homebrew/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 664, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 627, in _load_backward_compatible
  File "/opt/homebrew/opt/pygobject3/lib/python3.9/site-packages/gi/importer.py", line 144, in load_module
    importlib.import_module('gi.repository.' + dep.split("-")[0])
  File "/opt/homebrew/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 664, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 627, in _load_backward_compatible
  File "/opt/homebrew/opt/pygobject3/lib/python3.9/site-packages/gi/importer.py", line 145, in load_module
    dynamic_module = load_overrides(introspection_module)
  File "/opt/homebrew/opt/pygobject3/lib/python3.9/site-packages/gi/overrides/__init__.py", line 118, in load_overrides
    override_mod = importlib.import_module(override_package_name)
  File "/opt/homebrew/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "/opt/homebrew/opt/pygobject3/lib/python3.9/site-packages/gi/overrides/Pango.py", line 41, in <module>
    FontDescription = override(FontDescription)
  File "/opt/homebrew/opt/pygobject3/lib/python3.9/site-packages/gi/overrides/__init__.py", line 195, in override
    assert g_type != TYPE_NONE
AssertionError
```
Probably a version conflict with pango? 

No compilation reported compilation errors:

```
% brew info pympress       
pympress: stable 1.5.3, HEAD
Simple and powerful dual-screen PDF reader designed for presentations
https://github.com/Cimbali/pympress/
/opt/homebrew/Cellar/pympress/1.5.3_2 (1,296 files, 14.7MB) *
  Built from source on 2021-04-11 at 20:29:35
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/pympress.rb
License: GPL-2.0
==> Dependencies
Required: gobject-introspection ✔, gtk+3 ✔, libyaml ✔, poppler ✔, pygobject3 ✔, python@3.9 ✔
==> Options
--HEAD
	Install HEAD version
==> Analytics
install: 60 (30 days), 255 (90 days), 905 (365 days)
install-on-request: 60 (30 days), 256 (90 days), 906 (365 days)
build-error: 0 (30 days)
```

What is reported in pympress.log?\*

Does the problem still happen if you remove your config file?\*
config file non-existent

**Comment by Cimbali** [comment]
Alright, it does seem like the problem is the pango library:

`Failed to load shared library 'libpango-1.0.0.dylib' referenced by the typelib: dlopen(libpango-1.0.0.dylib, 9): image not found`

It seems like the recipe does not depend on pango/cairo, which is probably a mistake (but I’m surprised this didn’t cause problems before, maybe it wasn’t always a separate package, or the library also shipped in gtk+3?).

Can you try `brew install cairo pango` to see if that fixes the issue ? If so I’ll just adjust the homebrew dependencies.

**Comment by LightServant** [comment]
Hi!

I did have pango installed but not cairo.
Did a `brew install cairo` as requested, followed by `brew install --build-from-source pympress` to recompile.

Unfortunately the error remains the same.
I'm attaching a zip-file with the brew compilation logs from `~/Library/Logs/Homebrew`
I don't see anything out of the order there, but maybe you can spot something. 

If you need more information, let me know. 

[pympress.zip](https://github.com/Cimbali/pympress/files/6292867/pympress.zip)

**Comment by Cimbali** [comment]
This means most likely there is an error in the build process of gtk out

**Comment by Cimbali** [comment]
The similar issue was #58

Can you try the following and report the output, and whether the listed files exist?

```
ings $(brew --prefix)/lib/girepository-1.0/*.typelib | grep libpango-[0-9.]*.dylib 
```

**Comment by LightServant** [comment]
You're onto something:

```
% strings $(brew --prefix)/lib/girepository-1.0/*.typelib | grep libpango-[0-9.]*.dylib
zsh: no matches found: libpango-[0-9.]*.dylib
```

So what's the fix?  I don't see the solution exactly in #58.
This is a bug in girepository then or in pango, for not registering with girepository?

**Comment by Cimbali** [comment]
I think [this](https://gitlab.gnome.org/GNOME/gobject-introspection/-/merge_requests/11/) was the fix ? Not entirely sure, but it was upstream, not in pympress.

**Comment by Cimbali** [comment]
> theres not a single file linked to a *.dylib extension

The files don’t link to dylibs, but should contain the paths, hence the `strings` command. If you just grep for `dylib` on the strings output you should see a number of them.

**Comment by LightServant** [comment]
Well then I think you can close and I need to reopen on gi-introspection, correct?

**Comment by Cimbali** [comment]
You should probably open a bug there, especially if you can reproduce it with a minimal example. I can tag it here with `external dependency` so we can track it.


If you run a python file that just imports Pango from gobject introspections for example, does it fail?

```python
import gi
gi.require_version('Pango', '1.0')
from gi.repository import Pango
print(Pango.version_string())
```

**Comment by DocEpsilon** [comment]
> You should probably open a bug there, especially if you can reproduce it with a minimal example. I can tag it here with `external dependency` so we can track it.

Done: https://gitlab.gnome.org/GNOME/gobject-introspection/-/issues/383
They already forwarded me to the homebrew issue tracker. I'll drop the issue number here once I'm there.

In the meantime, thanks, for your exceptional support! Due to the bug, I haven't been able to evaluate your software yet. But I'm sure if it's half as good as the support I'll become a user and will make lots of advertisement and make a small donation. :-)

**Comment by DocEpsilon** [comment]
fyi: Executing a workaround like `%  DYLD_LIBRARY_PATH=/opt/homebrew/lib pympress presentation.pdf `
fails with `segmentation fault `

**Comment by Cimbali** [comment]
Not the best news :) this means we don’t have a functional workaround.
DYLD_LIBRARY_PATH tends to mess with a lot of other components looking up their libraries, so we’ll just have to wait or homebrew’s fix.

**Comment by DocEpsilon** [comment]
Pull request for pango is issued on hombrew-core:
https://github.com/Homebrew/homebrew-core/pull/75484

However, with my local changes I still can't start pympress. 

It now fails with:
```% pympress presentation.pdf 

** (process:88315): WARNING **: 14:31:45.218: Failed to load shared library 'libgdk_pixbuf-2.0.0.dylib' referenced by the typelib: dlopen(libgdk_pixbuf-2.0.0.dylib, 9): image not found
```

:( 

New issue or follow this one?  Currently I'm unaware which package/bottel `libgdk_pixbuf-2.0.0.dylib` belongs to.

**Comment by Cimbali** [comment]
Sounds like [gdk-pixbuf](https://formulae.brew.sh/formula/gdk-pixbuf) to me, we can keep tracking the issue here. Hopefully the same fix works.

**Comment by DocEpsilon** [comment]
Unfortunatly the fix did not pass review. At least up to now there has been a veto:
https://github.com/Homebrew/homebrew-core/pull/75484#pullrequestreview-638842663
Maybe you'd like to comment?

**Comment by ilippert** [comment]
python3-pympress-1.6.0-1.noarch

On Linux Fedora I get 
`Traceback (most recent call last):
  File "/usr/bin/pympress", line 33, in <module>
    sys.exit(load_entry_point('pympress==1.6.0', 'gui_scripts', 'pympress')())
  File "/usr/bin/pympress", line 25, in importlib_load_entry_point
    return next(matches).load()
StopIteration
`
then pympress aborts.


Is this the same issue, or should this be a new issue?

**Comment by Cimbali** [comment]
While waiting for Homebrew/homebrew-core#76383 to be merged to fix the library search paths, let’s see if removing some features (one that uses TreeStore, one that caused problems in the past and similarly accesses some buffers) helps.

Could you try installing pympress with the command below ?

    pip install "git+https://github.com/Cimbali/pympress@dumbed-down-interface#egg=pympress"

(Optionally with `--upgrade` if pip complains pympress is already installed).

If pympress still segfaults after that I’m not sure which component (if any specifically) could be causing the segfault, so we’d have to do some more digging into the origin of the error.

**Comment by DocEpsilon** [comment]
I've installed/upgraded as you requested. Pympress unfortunately still segfaults: 
https://gist.github.com/DocEpsilon/ddf30456ad350e7c2ff5497117676726

**Comment by Cimbali** [comment]
Really similar error too. Well I'm afraid we have to know more about what is causing this segfault before we can do anything else.

**Comment by Cimbali** [comment]
Apparently [faulthandler](https://docs.python.org/3/library/faulthandler.html) can help identify inside which python functions the segfault happens. Maybe running it a couple of times can help identify if there are several different functions causing issues.

**Comment by DocEpsilon** [comment]
Does this output help?
```
% python3 -X faulthandler -m pympress presentation.pdf
Fatal Python error: Segmentation fault

Current thread 0x000000010082bd40 (most recent call first):
  File "/opt/homebrew/lib/python3.9/site-packages/pympress/builder.py", line 158 in connect_signals
  File "/opt/homebrew/lib/python3.9/site-packages/pympress/scribble.py", line 170 in __init__
  File "/opt/homebrew/lib/python3.9/site-packages/pympress/ui.py", line 233 in __init__
  File "/opt/homebrew/lib/python3.9/site-packages/pympress/app.py", line 110 in __init__
  File "/opt/homebrew/lib/python3.9/site-packages/pympress/__main__.py", line 105 in main
  File "/opt/homebrew/lib/python3.9/site-packages/pympress/__main__.py", line 109 in <module>
  File "/opt/homebrew/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/runpy.py", line 87 in _run_code
  File "/opt/homebrew/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/runpy.py", line 197 in _run_module_as_main
zsh: segmentation fault  python3 -X faulthandler -m pympress presentation.pdf
```

**Comment by Cimbali** [comment]
It is something, indeed. I removed he highlight feature too, since it seems that the initialization of that feature is where the error happens. I pushed it to the dumbed-down-interface branch, so you can update with the previous pip command.

**Comment by DocEpsilon** [comment]
Hmm, somethings wrong. pympress won't install from the command above, even with `--upgrade`-option set. So I did a `pip uninstall pympres` follow by `pip install "git+https://github.com/Cimbali/pympress@dumbed-down-interface#egg=pympress"`which still segfaults... 
Are you sure you pushed sucessfully to the remote branch ?

**Comment by Cimbali** [comment]
Same function but called from somewhere else… I’m afraid we can’t try to take away “connect signals” − nothing would work anymore. I think all we can do now is weigh in on [the pygobject issue](https://gitlab.gnome.org/GNOME/pygobject/-/issues/455), maybe provide them with a simple example to reproduce the error. If they don’t have guidance or pointers on how to fix this I’m not sure who does.

**Comment by Cimbali** [comment]
Hi @DocEpsilon, the fix seems to have been merged in the upstream gobject-introspection code now. You have 2 options:
- wait until gobject-introspection 1.71.0 is released and makes it into Homebrew, then try with that
- try to get and compile the gobject-introspection current master

If you want to go with the latter, I think you can edit the formula with `brew edit gobject-introspection` to add
```rb
  head "https://gitlab.gnome.org/GNOME/gobject-introspection.git", branch: "master"
```
And then install from HEAD:
```sh
brew install --HEAD --build-from-source gobject-introspection
```

**Comment by Cimbali** [comment]
Currently on 1.74 and 1.72 was released in March, so I’ll consider this solved.
